### PR TITLE
Use Diskuv OCaml as the current Windows option

### PR DESCRIPTION
**Pull request by jonahbeckford**

### CODE REVIEW

**Comment by jonahbeckford** [comment]
Note: The failing CI Build has nothing to do with the PR!

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (tmattio) [inline data/tutorials/gs_03_ocaml_on_windows.md]: `opam.2.2.0` isn't an installation environment, we could actually move this to the introduction section.

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (tmattio) [inline data/tutorials/gs_03_ocaml_on_windows.md]: I don't think that's true anymore, c.f. https://github.com/ocamllabs/vscode-ocaml-platform/pull/481

**Comment by Leonidas-from-XIV** [comment]
Since #543 was merged, the CI is fixed so if you rebase on `main` the PR it should finish successfully :)

**Review by tmattio — COMMENTED** [review]
Thanks a lot Jonah!

I see that you updated the Up and Running documentation. All the changes look good (my suggestions on the recommended Windows method aside, more on that below), but we recently rewrote that page in anticipation to the release of OCaml 5: https://staging.ocaml.org/docs/up-and-running. I don't mind merging changes to the Up and Running documentation now, but would you like to also port your changes to the staging branch?

Re. the new Windows documentation, I'm not sure about making Diskuv the default installation method.

Granted, it is simpler to install than Cygwin + fdopen's repository, but I think the recommended method should support the latest compiler version, especially with the upcoming OCaml 5 release (it'll be a bit awkward when we advertise OCaml 5 that our official documentation recommends not to use it).

It'd be great to point to Diskuv as a Windows solution, but perhaps as an alternative if one is willing to deviate from the official compiler release cycle?

The decision process, IMO, would look like this:

- When you want to develop applications, use Cygwin + fdopen's repo
- When you want to develop applications, care about stability, and don't care about the compiler version, use Diskuv.
- When you only want to run applications, use WSL or Docker

What do you think?

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (jonahbeckford) [reply data/tutorials/gs_03_ocaml_on_windows.md]: I can remove this.

**Comment by jonahbeckford** [comment]
> It'd be great to point to Diskuv as a Windows solution, but perhaps as an alternative if one is willing to deviate from the official compiler release cycle?
> 
> The decision process, IMO, would look like this:
> 
> * When you want to develop applications, use Cygwin + fdopen's repo
> * When you want to develop applications, care about stability, and don't care about the compiler version, use Diskuv.
> * When you only want to run applications, use WSL or Docker
> 
> What do you think?

Agree. And I'll copy-edit the rest so your good point about not tracking the official compiler release cycle is emphasized.

*Edit*: I missed over the fdopen repository part of Cygwin + fdopen; I mixed that up with Cygwin + central Opam with appropriate `available: [ os != win32 ]` flags. But it doesn't seem responsible to recommend developing an application with a repository where, on the same page, we say (with added italics):
> For a long time this *has been* maintained by [@fdopen](https://fdopen.github.io/opam-repository-mingw/) ... As of August 2021, [the repository *will no longer* be updated](https://fdopen.github.io/opam-repository-mingw/2021/02/26/repo-discontinued/)

IMHO, recommending something is equivalent to supporting the user when they file a ticket. From what I understand the only place for a Windows user today to get a response for a Windows Opam package related issue is to file an issue with Diskuv's distribution. User support has been my implicit dividing line between "primary" recommendations and "alternatives". There could be some other important dividing lines, and I think you are alluding to OCaml 5 support being one dividing line (which makes sense).

Perhaps we need multiple dividing lines like the Tier *N* labeling of the OCaml compiler: Tier 1 ("Actively Maintained"; Ticket Supported + Latest Compilers), Tier 2 ("Maintained when possible"; Ticket Supported) and Tier 3 (User Supported). Tier 1 would be what you are calling Opam 2.2. Tier 2 would be Diskuv. Tier 3 would be fdopen, Docker and WSL2.

Does that grouping work better?

**Comment by jonahbeckford** [comment]
> I don't mind merging changes to the Up and Running documentation now, but would you like to also port your changes to the staging branch?

Yes, I'll merge the changes to staging branch as well.

**Comment by jonahbeckford** [comment]
Ported to staging branch with same content but re-organized in a way that de-emphasizes Diskuv OCaml (platform installer + Opam 2.2 should be used instead). #554 

Incorporated feedback.

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
**Guidelines for Following Instructions on this Page**
```
Title case to match the other subheaders.

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
A **shell** is a program that will let you enter commands in a text windows using only your keyboard. It's also known as a command line interface (CLI). When this page asks you to enter commands in your shell, use the following instructions for your system:
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:24]:
```suggestion
* On macOS, you will run the `Terminal` app to start a shell.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* On Windows, you can start PowerShell by pressing the Windows key (`⊞`), typing "PowerShell" and then clicking Open `Windows PowerShell`. There is an older shell called "Command Prompt" you can use as well.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* On Linux, you are already familiar with a shell (typically bash or zsh).
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
The foundation for installing OCaml and third-party OCaml libraries is [opam](https://opam.ocaml.org/), OCaml's package manager. On Unix-like systems, you'll install and manage opam through the Unix shell; on Window, you'll use a traditional `setup.exe` installer that will install and manage opam on your behalf. Whether you are on Unix or Windows, you'll end up with OCaml ready for you to use.
```
Thomas confirmed the official name of opam is all lowercase.

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
In the code blocks below, `# Homebrew` indicates the command if you're using [Homebrew](https://brew.sh/) as the installer, and `# MacPorts` if you're using [MacPorts](https://www.macports.org/) to install.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
It's essential to install the OCaml compiler because OCaml is a compiled language, so please don't skip this step. The first part sets up opam, and the second part installs the OCaml base compiler:
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
In this section, we'll describe using the new [Diskuv OCaml](https://github.com/diskuv/dkml-installer-ocaml#readme) ("DKML") Windows installer. Expect to see another officially-supported Windows installation provided directly by opam in the coming months; it will be compatible with your DKML installation.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* You need to **stay at your computer** and press "Yes" for any Windows security popups. 
After the DKML installer finishes installing two programs (`Visual Studio Installer`
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
Now, download and run:
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
All the build outputs generated by Dune go in the `_build` directory. The
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
* **only to run, not develop, applications**, use [Docker](#docker-images) or [WSL2](#wsl2)
* **to develop applications and have some familiarity with Unix**, use [opam-repository-mingw](#opam-repository-mingw)
* **to develop applications and care more about stability and ease-of-use than the latest compiler**, use [Diskuv OCaml](/docs/up-and-running)
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
### `opam-repository-mingw`
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
**If you use the recommended DKML installer**, you will need to:
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
**If you use WSL2**, you will remotely connect to your WSL2 instance from
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
**If you use Vim**, the default Cygwin Vim will not work with
Merlin. You will need install Vim separately. In addition to the usual
instructions printed when installing Merlin, you may need to set the PATH in
```

**Review by christinerose — CHANGES_REQUESTED** [review]
Mostly line editing for clarity, grammar, consistency, and syntax. There will be some conflicts with the recently merged PR, as I had also gone through this, and I think I incorporated some of these suggestions in explaining the shell in Up and Running.

**Comment by tmattio** [comment]
@jonahbeckford seeing the number of PRs made on the Up and Running tutorial, we decided to merge the new version that was on staging immediately to avoid having to rebase constantly. I can rebase your current PR on top of main if you want. A lot of changes in the new version are actually extracted from your PR, with some edits from Christine.

**Comment by jonahbeckford** [comment]
I can/will rebase.

**Comment by jonahbeckford** [comment]
I rebased. I think I got all the suggestions merged, but now there is also a bit of new content.

File: data/tutorials/gs_00_up_and_running.md
Comment (jonahbeckford) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* OCaml 4.12.1 with Git and Visual Studio compiler: [setup-diskuv-ocaml-windows_x86_64-1.0.1.exe](https://github.com/diskuv/dkml-installer-ocaml/releases/download/v1.0.1/setup-diskuv-ocaml-windows_x86_64-1.0.1.exe)
```

**Review by jonahbeckford — COMMENTED** [review]
Update to Diskuv OCaml v1.0.1

**Comment by tmattio** [comment]
Thanks for rebasing @jonahbeckford! I had another review pass and it looks really good to me.

@christinerose @panglesd do you want to have a look at the version after the rebase? Happy to merge when it looks good to you too.

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
A **shell** is a program that will let you enter commands in a text window using only your keyboard. It's also known as a command line interface (CLI). When this page asks you to enter commands in your shell, use the following instructions for your system:
```

It will be "a text window" or "text windows" (without the article "a")

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* On Windows, you can start PowerShell by pressing the Windows key (`⊞`), typing "PowerShell," and then clicking Open `Windows PowerShell`. There is an older shell called "Command Prompt" you can use as well
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
On Unix, the best way to install OCaml is with opam, OCaml's package manager. On Windows, the best way is to use a traditional `setup.exe` that will initialise opam
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
on your behalf; however, only OCaml version 4.12.1 can be installed with `setup.exe` now.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
[opam](https://opam.ocaml.org/), OCaml's package manager, introduces the concept of "switches," which is a compiler with a set of packages (libraries and other files). Switches are used to have independent sets of dependencies in different projects.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* Linux or macOS: [Installation for Unix, including Linux and macOS](#installation-for-unix)
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
On Unix, it's essential to initialise opam, which will (if needed) install the OCaml compiler. If you already have OCaml installed, opam will use that compiler.  
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
> Advanced Users: If you are familiar with Cygwin or WSL2, there are other installation methods described on the [OCaml on Windows](/docs/ocaml-on-windows) page.
```

Since there weren't quotes around these at the top of this document, I removed them here.

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
  and `Git for Windows`), you can leave your computer for the remaining two (2) hours.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
OCaml comes with two additional compilers: one compiles to **native code** (sometimes called machine code or executable binary), directly read by the CPU, and the other compiles to **bytecode**, creating an executable that can be interpreted by a variety of runtime environments, making more it flexible.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
**For Vim and Emacs**, we won't use the LSP server but rather directly talk to Merlin.
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* On macOS and Linux, use the directory listing command (`ls`):
```

File: data/tutorials/gs_00_up_and_running.md
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md]:
```suggestion
* On Windows, use the directory listing command (`dir`):
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
platform with a [major release of opam](#opam-22) in the coming months, and it will be compatible with
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md:22]: There are so many commas in this sentence.
 Perhaps we can reword it?
Like: 
**only to run (not develop) applications**, use [Docker](#docker-images) or [WSL2](#wsl2)

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
opam will focus on closing the gap to fully supporting Windows. This includes
```

File: data/tutorials/gs_03_ocaml_on_windows.md
Comment (christinerose) [inline data/tutorials/gs_03_ocaml_on_windows.md]:
```suggestion
supporting an external dependency installation for Windows and integrating it with the
```

**Review by christinerose — CHANGES_REQUESTED** [review]

**Comment by tmattio** [comment]
Thanks a lot @jonahbeckford, it's a massive improvement over the previous Windows setup documentation! :tada:

**Pull request merged**
